src/main/java/com/metamx/common/FileUtils.java

```diff
@@ -0,0 +1,302 @@
+   *
+   * @return `true` if the digests match, `false` if they do not
+   *
+   * @throws IOException passed from inner stream errors
+   */
+  public static Boolean checkDigest(final File file, final MessageDigest otherDigest) throws IOException
+  {
+    return FileCopyResult.buildResult(file.toPath(), otherDigest.getAlgorithm()).digestEquals(otherDigest);
+  }
+
+  /**
+   * Copy input stream to outFile. If outFile exists, it is attempted to be deleted. Also validates the outFile with MD5
+   *
+   * @param byteSource  Supplier for an input stream that is to be copied. The resulting stream is closed each iteration
+   * @param outFile     Where the file should be written to.
+   * @param shouldRetry Predicate indicating if an error is recoverable and should be retried. Note that MD5MismatchException may be thrown by the inner loop in addition to IOException
+   * @param maxAttempts The maximum number of assumed recoverable attempts to try before completely failing.
+   *
+   * @throws java.lang.RuntimeException wrapping the inner exception on failure.
+   */
+  public static FileCopyResult retryCopy(
+      final ByteSource byteSource,
+      final File outFile,
+      final Predicate<Throwable> shouldRetry,
+      final int maxAttempts
+  )
+  {
+    try {
+      return RetryUtils.retry(
+          new Callable<FileCopyResult>()
+          {
+            @Override
+            public FileCopyResult call() throws Exception
+            {
+              try (final InputStream inputStream = byteSource.openStream()) {
+                if (outFile.exists()) {
+                  if (!outFile.delete()) {
+                    throw new IAE("Could not remove file at [%s]", outFile.getAbsolutePath());
+                  }
+                }
+
+                final MessageDigest digest = StreamUtils.copyToFileWithInputDigest(inputStream, outFile, "MD5");
+                final FileCopyResult result = FileCopyResult.buildResult(outFile.toPath(), "MD5");
+                if (!result.digestEquals(digest)) {
+                  throw new MD5MismatchException(
+                      "File at [%s] did not match MD5 after copy",
+                      outFile.getAbsolutePath()
+                  );
+                }
+                return result;
+              }
+            }
+          },
+          shouldRetry,
+          maxAttempts
+      );
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Retry copy attempts from input stream to output stream. Does *not* check to make sure data was intact during the transfer
+   *
+   * @param byteSource  Supplier for input streams to copy from. The stream is closed on every retry.
+   * @param byteSink    Supplier for output streams. The stream is closed on every retry.
+   * @param shouldRetry Predicate to determine if the throwable is recoverable for a retry
+   * @param maxAttempts Maximum number of retries before failing
+   */
+  public static long retryCopy(
```

Comment:
I feel it's confusing to have to methods with the same name, with very similar signatures to have very different behavior. The behavior should be more or less clear from the name. It's also confusing why one would take a ByteSink and the other one a File, one return a long, the other one a FileCopyResult. I feel this should be two different methods.